zipkin/src/main/java/zipkin2/codec/SpanBytesDecoder.java

```diff
@@ -35,123 +37,180 @@
     }
 
-    @Override
-    public boolean decodeList(byte[] spans, Collection<Span> out) { // ex getTrace
-      return JsonCodec.readList(new V2SpanReader(), spans, out);
+    @Override public boolean decodeList(ByteBuffer spans, Collection<Span> out) {
+      return JsonCodec.readList(new V2SpanReader(), ReadBuffer.wrapUnsafe(spans), out);
     }
 
-    @Override
-    @Nullable
-    public Span decodeOne(byte[] span) {
-      return JsonCodec.readOne(new V2SpanReader(), span);
+    @Override @Nullable public Span decodeOne(byte[] span) {
+      return JsonCodec.readOne(new V2SpanReader(), ReadBuffer.wrap(span, 0));
     }
 
-    @Override
-    public List<Span> decodeList(byte[] spans) {
+    @Override @Nullable public Span decodeOne(ByteBuffer span) {
+      return JsonCodec.readOne(new V2SpanReader(), ReadBuffer.wrapUnsafe(span));
+    }
+
+    @Override public List<Span> decodeList(byte[] spans) {
+      return decodeList(this, spans);
+    }
+
+    @Override public List<Span> decodeList(ByteBuffer spans) {
       return decodeList(this, spans);
     }
   },
   PROTO3 {
-    @Override
-    public Encoding encoding() {
+    @Override public Encoding encoding() {
       return Encoding.PROTO3;
     }
 
-    @Override
-    public boolean decode(byte[] span, Collection<Span> out) { // ex decode span in dependencies job
-      return Proto3Codec.read(span, out);
+    @Override public boolean decode(byte[] span, Collection<Span> out) { // ex DependencyLinker
+      return Proto3Codec.read(ReadBuffer.wrap(span, 0), out);
+    }
+
+    @Override public boolean decodeList(byte[] spans, Collection<Span> out) { // ex getTrace
+      return Proto3Codec.readList(ReadBuffer.wrap(spans, 0), out);
+    }
+
+    @Override public boolean decodeList(ByteBuffer spans, Collection<Span> out) {
+      return Proto3Codec.readList(ReadBuffer.wrapUnsafe(spans), out);
     }
 
-    @Override
-    public boolean decodeList(byte[] spans, Collection<Span> out) { // ex getTrace
-      return Proto3Codec.readList(spans, out);
+    @Override @Nullable public Span decodeOne(byte[] span) {
+      return Proto3Codec.readOne(ReadBuffer.wrap(span, 0));
     }
 
-    @Override
-    @Nullable
-    public Span decodeOne(byte[] span) {
-      return Proto3Codec.readOne(span);
+    @Override @Nullable public Span decodeOne(ByteBuffer span) {
+      return Proto3Codec.readOne(ReadBuffer.wrapUnsafe(span));
     }
 
-    @Override
-    public List<Span> decodeList(byte[] spans) {
+    @Override public List<Span> decodeList(byte[] spans) {
+      return decodeList(this, spans);
+    }
+
+    @Override public List<Span> decodeList(ByteBuffer spans) {
       return decodeList(this, spans);
     }
   };
 
+  /**
+   * ByteBuffer implementation of {@link #decodeList(byte[])}.
+   *
+   * <p>Note: only use this when it is ok to modify the underlying {@link ByteBuffer#array()}.
+   */
+  public abstract boolean decodeList(ByteBuffer spans, Collection<Span> out);
```

Comment:
adding the methods here allows us to not break the BytesDecoder interface, yet still allow us the overloads for our builtin types